src/samples/coap/lwm2m-client.c

```diff
@@ -707,6 +726,276 @@ del_location_obj(void *instance_data, void *user_data,
+    struct sol_lwm2m_tlv *tlv;
+    struct access_control_obj_instance_ctx *instance_ctx = instance_data;
+    struct acl_instance *acl_item;
+
+    SOL_VECTOR_FOREACH_IDX (tlvs, tlv, i) {
+        if (tlv->id == ACCESS_CONTROL_OBJ_OBJECT_RES_ID &&
+            tlv->type == SOL_LWM2M_TLV_TYPE_RESOURCE_WITH_VALUE) {
+            r = sol_lwm2m_tlv_get_int(tlv, &instance_ctx->object_id);
+            if (r < 0)
+                return r;
+        } else if (tlv->id == ACCESS_CONTROL_OBJ_INSTANCE_RES_ID &&
+            tlv->type == SOL_LWM2M_TLV_TYPE_RESOURCE_WITH_VALUE) {
+            r = sol_lwm2m_tlv_get_int(tlv, &instance_ctx->instance_id);
+            if (r < 0)
+                return r;
+        } else if (tlv->id == ACCESS_CONTROL_OBJ_ACL_RES_ID &&
+            tlv->type == SOL_LWM2M_TLV_TYPE_MULTIPLE_RESOURCES) {
+                uint16_t j = i + 1;
+                struct sol_lwm2m_tlv *res_tlv;
+                int64_t res_val;
+
+                sol_vector_clear(&instance_ctx->acl);
+
+                while ((res_tlv = sol_vector_get(tlvs, j)) &&
+                    res_tlv->type == SOL_LWM2M_TLV_TYPE_RESOURCE_INSTANCE) {
+                    r = sol_lwm2m_tlv_get_int(res_tlv, &res_val);
+                    if (r < 0)
+                        goto err_free_acl;
+
+                    acl_item = sol_vector_append(&instance_ctx->acl);
+                    if (!acl_item) {
+                        fprintf(stderr, "Could not alloc memory for access control list Resource Instance\n");
+                        r = -ENOMEM;
+                        goto err_free_acl;
+                    }
+
+                    acl_item->key = res_tlv->id;
+                    acl_item->value = res_val;
+                    fprintf(stderr, "<<[WRITE_TLV]<< acl[%" PRIu16 "]=%" PRId64 ">>>>\n", acl_item->key, acl_item->value);
+                    j++;
+                }
+                i = j - 1;
+        } else if (tlv->id == ACCESS_CONTROL_OBJ_OWNER_RES_ID &&
+            tlv->type == SOL_LWM2M_TLV_TYPE_RESOURCE_WITH_VALUE) {
+            r = sol_lwm2m_tlv_get_int(tlv, &instance_ctx->owner_id);
+            if (r < 0)
+                return r;
+        } else {
+            fprintf(stderr, "tlv type: %u, ID: %" PRIu16 ", Size: %zu, Content: %.*s"
+                " could not be written to Access Control Object at /2/%" PRIu16,
+                tlv->type, tlv->id, tlv->content.used,
+                SOL_STR_SLICE_PRINT(sol_buffer_get_slice(&tlv->content)), instance_id);
+            return -EINVAL;
+        }
+    }
+
+    if (tlvs->len == 1 && r >= 0)
+        printf("TLV written to Access Control object at /2/%" PRIu16 "/%" PRIu16 "\n",
+            instance_id, tlv->id);
+    else
+        printf("TLV written to Access Control object at /2/%" PRIu16 "\n", instance_id);
+
+    return r;
+
+err_free_acl:
+    sol_vector_clear(&instance_ctx->acl);
+    return r;
+}
+
+static int
+create_access_control_obj(void *user_data, struct sol_lwm2m_client *client,
```

Comment:
This function and  `write_access_control_tlv` shares common code patterns. You could create auxiliary functions instead of doing copy & pasta.